tools/SourceKit/lib/SwiftLang/SwiftEditor.cpp

```diff
@@ -2535,3 +2505,60 @@ void SwiftLangSupport::editorExpandPlaceholder(StringRef Name, unsigned Offset,
   }
   EditorDoc->expandPlaceholder(Offset, Length, Consumer);
 }
+
+//===----------------------------------------------------------------------===//
+// Semantic Tokens
+//===----------------------------------------------------------------------===//
+
+void SwiftLangSupport::getSemanticTokens(
+    StringRef PrimaryFilePath, ArrayRef<const char *> Args,
+    llvm::Optional<VFSOptions> VfsOptions,
+    SourceKitCancellationToken CancellationToken,
+    std::function<void(const RequestResult<SemanticTokensResult> &)> Receiver) {
+  std::string FileSystemError;
+  auto FileSystem = getFileSystem(VfsOptions, PrimaryFilePath, FileSystemError);
+  if (!FileSystem) {
+    Receiver(RequestResult<SemanticTokensResult>::fromError(FileSystemError));
+    return;
+  }
+
+  std::string InvocationError;
+  SwiftInvocationRef Invok = ASTMgr->getTypecheckInvocation(
+      Args, PrimaryFilePath, FileSystem, InvocationError);
+  if (!InvocationError.empty()) {
+    LOG_WARN_FUNC("error creating ASTInvocation: " << InvocationError);
+  }
+  if (!Invok) {
+    Receiver(RequestResult<SemanticTokensResult>::fromError(InvocationError));
+    return;
+  }
+
+  class SemanticTokensConsumer : public SwiftASTConsumer {
+    std::function<void(const RequestResult<SemanticTokensResult> &)> Receiver;
+
+  public:
+    SemanticTokensConsumer(
+        std::function<void(const RequestResult<SemanticTokensResult> &)>
+            Receiver)
+        : Receiver(Receiver) {}
+
+    void handlePrimaryAST(ASTUnitRef AstUnit) override {
+      auto &CompIns = AstUnit->getCompilerInstance();
+      unsigned BufferID = AstUnit->getPrimarySourceFile().getBufferID().value();
+      SemanticAnnotator Annotator(CompIns.getSourceMgr(), BufferID);
+      Annotator.walk(AstUnit->getPrimarySourceFile());
```

Comment:
This should be the input file (which needs to be passed in along with the primary file), see eg. `retrieveInputFile` (which you might need to extract so you can use it here). That will allow grabbing semantic tokens for eg. macro expansions as well.

Reply:
Oh, I just copied the code from the request to get diagnostics but apparently that’s incorrect as well, then. I updated both this request and the diagnostics request to use `retrieveSourceFile`.